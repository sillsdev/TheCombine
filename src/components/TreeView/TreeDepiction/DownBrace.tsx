import { ReactElement } from "react";

interface DownBraceProps {
  backgroundColor?: string;
  color?: string;
  height: number;
  width: number;
}

export default function DownBrace(props: DownBraceProps): ReactElement {
  const color = props.color ?? "#C6CACC";
  const height = 16;
  const wScaled = props.width * (height / props.height);
  const wEnd = 9;
  const wMid = 16;
  const wRequired = 2 * wEnd + wMid;
  const wBar = Math.max(0, wScaled - wRequired) / 2;
  const width = wRequired + 2 * wBar;

  return (
    <svg
      height={props.height}
      preserveAspectRatio="none" // In case wRequired > wScaled
      viewBox={`0 0 ${width} ${height}`}
      width={props.width}
      xmlns="http://www.w3.org/2000/svg"
    >
      {/* Background */}
      <rect
        fill={props.backgroundColor ?? "white"}
        height={height}
        width={width}
      />

      {/* Endcap */}
      <g fill={color}>
        <path d="M2 16C2.03757 10.5664 4.2842 8.95611 9 9.00001L9 7C2.08822 7.03228 0.71149 8.74186 0 16L2 16Z" />
      </g>

      {/* Bar */}
      <g fill={color} transform={`translate(${wEnd})`}>
        <rect height={2} width={wBar} y={7} />
      </g>

      {/* Middle */}
      <g fill={color} transform={`translate(${wEnd + wBar})`}>
        <path d="M10 0C10.0376 5.4336 11.2842 7.04389 16 6.99999L16 9C9.08822 8.96772 6.7115 7.25814 6 0L10 0Z" />
        <path d="M6 0C5.96243 5.4336 4.71577 7.04389 0 6.99999L0 9C6.91178 8.96772 9.2885 7.25814 10 0L6 0Z" />
      </g>

      {/* Bar */}
      <g fill={color} transform={`translate(${wEnd + wBar + wMid})`}>
        <rect height={2} width={wBar} y={7} />
      </g>

      {/* Endcap */}
      <g fill={color} transform={`translate(${wEnd + wBar + wMid + wBar})`}>
        <path d="M7 16C6.96243 10.5664 4.71577 8.95611 0 9.00001L0 7C6.91178 7.03228 8.28851 8.74186 9 16L7 16Z" />
      </g>
    </svg>
  );
}
