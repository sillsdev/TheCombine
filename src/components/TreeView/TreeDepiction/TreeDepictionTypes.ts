import { SemanticDomain, SemanticDomainTreeNode } from "api/models";

const MAX_COL_WIDTH = 50; // Max gap.
const MIN_COL_WIDTH = 12; // Multiply this by RATIO_TILE_TO_GAP for min tile width.
export const RATIO_TILE_TO_GAP = 7; // Must be odd.

export function getNumCols(numChildren: number): number {
  return numChildren * (RATIO_TILE_TO_GAP + 1) - 1;
}

export function getColWidth(numChildren: number, totalWidth: number): number {
  if (!numChildren) {
    return MAX_COL_WIDTH;
  }
  const colWidth = Math.floor(totalWidth / getNumCols(numChildren));
  return Math.min(Math.max(colWidth, MIN_COL_WIDTH), MAX_COL_WIDTH);
}

export enum Direction {
  Down,
  Next,
  Prev,
  Up,
}

export interface TreeDepictionProps {
  animate: (domain: SemanticDomain) => Promise<void>;
  currentDomain: SemanticDomainTreeNode;
  small?: boolean;
}

export interface TreeRowProps extends TreeDepictionProps {
  colWidth: number;
}
