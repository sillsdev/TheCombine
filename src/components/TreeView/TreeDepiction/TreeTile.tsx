import { ImageListItem } from "@mui/material";
import { ReactElement } from "react";

interface TreeTileProps {
  colWidth: number;
  imgSrc: string;
}
/** Creates a section of the tree diagram (one of the branches) set to proper dimensions. */
export default function TreeTile(props: TreeTileProps): ReactElement {
  return (
    <ImageListItem style={{ transform: "scaleY(-1)" }}>
      <img
        src={props.imgSrc}
        alt={props.imgSrc}
        width={props.colWidth + 1} // The +1 is to avoid visual gaps from rounding errors
        height={60}
      />
    </ImageListItem>
  );
}
