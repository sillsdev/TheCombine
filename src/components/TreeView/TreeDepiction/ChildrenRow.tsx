import { ImageList, ImageListItem } from "@mui/material";
import { ReactElement } from "react";

import DomainTileButton from "components/TreeView/TreeDepiction/DomainTileButton";
import {
  Direction,
  RATIO_TILE_TO_GAP,
  TreeRowProps,
  getNumCols,
} from "components/TreeView/TreeDepiction/TreeDepictionTypes";
import TreeTile from "components/TreeView/TreeDepiction/TreeTile";
import * as tree from "resources/tree";

const HALF_TILE = (RATIO_TILE_TO_GAP - 1) / 2; // Half of cols-per-tile, rounded down

export default function ChildrenRow(props: TreeRowProps): ReactElement {
  // Creates a tile for the specified tree part
  const treeTile = (treeSrc: string): ReactElement => (
    <TreeTile
      colWidth={props.colWidth}
      imgSrc={treeSrc}
      key={treeSrc + Math.random()}
    />
  );

  // Creates a span across multiple columns
  const multiSpan = (cols: number): ReactElement[] => {
    return Array.from({ length: cols }, () => treeTile(tree.span));
  };

  // Creates empty items, half (rounded down) as many as there are columns in a tile
  const halfTileGap = (): ReactElement[] => {
    return Array.from({ length: HALF_TILE }, () => (
      <ImageListItem key={"gap" + Math.random()} />
    ));
  };

  // Creates the joist connecting current domain with subdomains
  const joistRow = (): ReactElement[] => {
    const row: ReactElement[] = [];
    const teeCount = props.currentDomain.children.length - 2;
    const middleTeeCount = teeCount % 2;
    const halfTeeCount = (teeCount - middleTeeCount) / 2;
    const rtl = document.body.dir === "rtl";

    // If there is only one child, the joist row has no branching
    if (teeCount === -1) {
      row.push(...halfTileGap());
      row.push(treeTile(tree.pillar));
      row.push(...halfTileGap());
      return row;
    }

    /* Start endcap */
    row.push(...halfTileGap());
    row.push(treeTile(rtl ? tree.endcapRight : tree.endcapLeft));

    /* Start -> Middle */
    for (let i = 0; i < halfTeeCount; i++) {
      row.push(...multiSpan(RATIO_TILE_TO_GAP));
      row.push(treeTile(rtl ? tree.teeUpLeft : tree.teeUpRight));
    }

    /* Middle */
    if (middleTeeCount) {
      row.push(...multiSpan(RATIO_TILE_TO_GAP));
      row.push(treeTile(tree.intersect));
      row.push(...multiSpan(RATIO_TILE_TO_GAP));
    } else {
      row.push(...multiSpan(HALF_TILE));
      row.push(treeTile(tree.teeDown));
      row.push(...multiSpan(HALF_TILE));
    }

    /* Middle -> End */
    for (let i = 0; i < halfTeeCount; i++) {
      row.push(treeTile(rtl ? tree.teeUpRight : tree.teeUpLeft));
      row.push(...multiSpan(RATIO_TILE_TO_GAP));
    }

    /* End endcap */
    row.push(treeTile(rtl ? tree.endcapLeft : tree.endcapRight));
    row.push(...halfTileGap());

    return row;
  };

  // Places the subdomain tiles
  const domainRow = (): ReactElement[] => {
    const subdomains: ReactElement[] = [];
    props.currentDomain.children.forEach((child, i) => {
      if (i > 0) {
        subdomains.push(<ImageListItem key={`GapTile${i}`} />);
      }
      subdomains.push(
        <ImageListItem key={`DomainTile${i}`} cols={RATIO_TILE_TO_GAP}>
          <DomainTileButton
            direction={Direction.Down}
            domain={child}
            onClick={(d) => props.animate(d)}
          />
        </ImageListItem>
      );
    });
    return subdomains;
  };

  const numCols = getNumCols(props.currentDomain.children.length);

  return (
    <ImageList
      cols={numCols}
      gap={0}
      rowHeight={"auto"}
      style={{ width: numCols * props.colWidth }}
    >
      {joistRow()}
      {domainRow()}
    </ImageList>
  );
}
