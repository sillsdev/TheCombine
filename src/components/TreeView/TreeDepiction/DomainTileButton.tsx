import {
  ChevronLeft,
  ChevronRight,
  KeyboardArrowDown,
  KeyboardArrowUp,
} from "@mui/icons-material";
import { Button, Grid, Typography } from "@mui/material";
import { CSSProperties, ReactElement } from "react";
import { useTranslation } from "react-i18next";

import { SemanticDomain } from "api/models";
import { Direction } from "components/TreeView/TreeDepiction/TreeDepictionTypes";

const RootId = "Sem";

interface DomainTextProps {
  domain: SemanticDomain;
  extraProps?: CSSProperties;
}

export function DomainText(props: DomainTextProps): ReactElement {
  const { t } = useTranslation();
  return (
    <div style={{ ...props.extraProps, textTransform: "capitalize" }}>
      <Typography variant={"overline"}>
        {props.domain.id !== RootId ? props.domain.id : ""}
      </Typography>
      <Typography variant={"body1"}>
        {props.domain.id !== RootId ? props.domain.name : t("addWords.domain")}
      </Typography>
    </div>
  );
}

interface DomainTileProps {
  direction: Direction;
  domain: SemanticDomain;
}

function DomainTile(props: DomainTileProps): ReactElement {
  const rtl = document.body.dir === "rtl";
  switch (props.direction) {
    case Direction.Down:
      return (
        <div>
          <DomainText domain={props.domain} />
          <KeyboardArrowDown />
        </div>
      );
    case Direction.Prev:
      return (
        <Grid
          container
          alignItems="center"
          justifyContent="space-around"
          wrap="nowrap"
        >
          <Grid item>{rtl ? <ChevronRight /> : <ChevronLeft />}</Grid>
          <Grid item>
            <DomainText domain={props.domain} />
          </Grid>
        </Grid>
      );
    case Direction.Next:
      return (
        <Grid
          container
          alignItems="center"
          justifyContent="space-around"
          wrap="nowrap"
        >
          <Grid item>
            <DomainText domain={props.domain} />
          </Grid>
          <Grid item>{rtl ? <ChevronLeft /> : <ChevronRight />}</Grid>
        </Grid>
      );
    case Direction.Up:
      return (
        <div>
          <KeyboardArrowUp />
          <DomainText domain={props.domain} />
        </div>
      );
  }
}

interface DomainTileButtonProps extends DomainTileProps {
  onClick: (domain: SemanticDomain) => void;
}

// Creates a semantic domain tile, which can be clicked on to navigate to that semantic domain
export default function DomainTileButton(
  props: DomainTileButtonProps
): ReactElement {
  return (
    <Button
      color="primary"
      id={props.domain.id}
      onClick={() => props.onClick(props.domain)}
      style={{
        left: 0,
        bottom: 0,
        width: "95%",
        height: "95%",
        margin: "2.5%",
        padding: "5px",
      }}
      variant={"outlined"}
    >
      <DomainTile {...props} />
    </Button>
  );
}
