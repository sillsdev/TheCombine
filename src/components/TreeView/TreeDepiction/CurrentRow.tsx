import { Button, ImageList, ImageListItem } from "@mui/material";
import { ReactElement } from "react";

import DomainTileButton, {
  DomainText,
} from "components/TreeView/TreeDepiction/DomainTileButton";
import {
  Direction,
  TreeDepictionProps,
} from "components/TreeView/TreeDepiction/TreeDepictionTypes";

export default function CurrentRow(props: TreeDepictionProps): ReactElement {
  const { next, previous } = props.currentDomain;

  const currentTile = (
    <Button
      color="primary"
      disabled={!props.currentDomain.parent}
      fullWidth
      id="current-domain"
      onClick={() => props.animate(props.currentDomain)}
      size="large"
      style={{ height: "95%" }}
      variant="contained"
    >
      <DomainText domain={props.currentDomain} extraProps={{ minWidth: 200 }} />
    </Button>
  );

  return props.small ? (
    currentTile
  ) : (
    <ImageList cols={7} gap={20} rowHeight={"auto"}>
      <ImageListItem cols={2}>
        {previous && (
          <DomainTileButton
            direction={Direction.Prev}
            domain={previous}
            onClick={props.animate}
          />
        )}
      </ImageListItem>
      <ImageListItem cols={3}>{currentTile}</ImageListItem>
      <ImageListItem cols={2}>
        {next && (
          <DomainTileButton
            direction={Direction.Next}
            domain={next}
            onClick={props.animate}
          />
        )}
      </ImageListItem>
    </ImageList>
  );
}
