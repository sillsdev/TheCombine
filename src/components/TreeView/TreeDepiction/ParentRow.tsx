import { ImageList, ImageListItem } from "@mui/material";
import { Fragment, ReactElement } from "react";

import DomainTileButton from "components/TreeView/TreeDepiction/DomainTileButton";
import {
  Direction,
  TreeRowProps,
} from "components/TreeView/TreeDepiction/TreeDepictionTypes";
import TreeTile from "components/TreeView/TreeDepiction/TreeTile";
import { parent } from "resources/tree";

export default function ParentRow(props: TreeRowProps): ReactElement {
  if (!props.currentDomain.parent) {
    return <Fragment />;
  }
  return (
    <ImageList cols={1} gap={0} rowHeight="auto">
      <ImageListItem>
        <DomainTileButton
          direction={Direction.Up}
          domain={props.currentDomain.parent}
          onClick={props.animate}
        />
      </ImageListItem>
      <TreeTile colWidth={props.colWidth} imgSrc={parent} />
    </ImageList>
  );
}
