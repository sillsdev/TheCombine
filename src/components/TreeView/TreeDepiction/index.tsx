import { Grid, Theme, useMediaQuery } from "@mui/material";
import { ReactElement, useEffect, useState } from "react";

import ChildrenRow from "components/TreeView/TreeDepiction/ChildrenRow";
import CurrentRow from "components/TreeView/TreeDepiction/CurrentRow";
import DomainTileButton from "components/TreeView/TreeDepiction/DomainTileButton";
import {
  Direction,
  TreeDepictionProps,
  getColWidth,
} from "components/TreeView/TreeDepiction/TreeDepictionTypes";
import { parent } from "resources/tree";
import { useWindowSize } from "utilities/useWindowSize";

export default function TreeDepiction(props: TreeDepictionProps): ReactElement {
  const showTree = useMediaQuery<Theme>((th) => th.breakpoints.up("sm"));

  const [colWidth, setColWidth] = useState(0);

  const { windowWidth } = useWindowSize();

  useEffect(() => {
    setColWidth(getColWidth(props.currentDomain.children.length, windowWidth));
  }, [props, windowWidth]);

  const currentDomain = props.currentDomain;
  return (
    <>
      {/* Display parent domain, if available. */}
      {showTree && currentDomain.parent && (
        <>
          <Grid item>
            <DomainTileButton
              direction={Direction.Up}
              domain={currentDomain.parent}
              onClick={props.animate}
            />
          </Grid>
          <Grid item>
            <img
              src={parent}
              style={{ transform: "scaleY(-1)" }}
              width={colWidth}
            />
          </Grid>
        </>
      )}

      {/* Display current domain and (if available) left and right brothers. */}
      <Grid item>
        <CurrentRow {...props} small={!showTree} />
      </Grid>

      {/* Display subdomains, if available. */}
      <Grid item>
        {showTree && currentDomain.children.length > 0 && (
          <ChildrenRow {...props} colWidth={colWidth} />
        )}
      </Grid>
    </>
  );
}
