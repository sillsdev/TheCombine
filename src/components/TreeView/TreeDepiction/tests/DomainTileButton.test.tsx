import { Button } from "@mui/material";
import { ReactTestRenderer, act, create } from "react-test-renderer";

import DomainTileButton from "components/TreeView/TreeDepiction/DomainTileButton";
import { Direction } from "components/TreeView/TreeDepiction/TreeDepictionTypes";
import domMap, { mapIds } from "components/TreeView/tests/SemanticDomainMock";

let tileMaster: ReactTestRenderer;
const MOCK_ANIMATE = jest.fn();

describe("DomainTileButton", () => {
  it("renders tile and matches the latest snapshot", async () => {
    await createTile();
    snapTest();
  });

  it("calls function on click", async () => {
    await createTile();
    expect(MOCK_ANIMATE).toHaveBeenCalledTimes(0);
    tileMaster.root.findByType(Button).props.onClick();
    expect(MOCK_ANIMATE).toHaveBeenCalledTimes(1);
  });
});

// Creates the tile
async function createTile(direction = Direction.Next): Promise<void> {
  await act(async () => {
    tileMaster = create(
      <DomainTileButton
        direction={direction}
        domain={domMap[mapIds.parent]}
        onClick={MOCK_ANIMATE}
      />
    );
  });
}

// Perform a snapshot test
function snapTest(): void {
  expect(tileMaster.toJSON()).toMatchSnapshot();
}
