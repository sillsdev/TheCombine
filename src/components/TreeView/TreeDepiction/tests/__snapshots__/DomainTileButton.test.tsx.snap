// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DomainTileButton renders tile and matches the latest snapshot 1`] = `
<button
  className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-1blvszr-MuiButtonBase-root-MuiButton-root"
  disabled={null}
  id="1"
  onBlur={[Function]}
  onClick={[Function]}
  onContextMenu={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  style={
    Object {
      "bottom": 0,
      "height": "95%",
      "left": 0,
      "margin": "2.5%",
      "padding": "5px",
      "width": "95%",
    }
  }
  tabIndex={-1}
  type="button"
>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-wrap-xs-nowrap css-mmuq0c-MuiGrid-root"
  >
    <div
      className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <div
        style={
          Object {
            "textTransform": "capitalize",
          }
        }
      >
        <span
          className="MuiTypography-root MuiTypography-overline css-1c7xfgu-MuiTypography-root"
          style={Object {}}
        >
          1
        </span>
        <p
          className="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
          style={Object {}}
        >
          parent
        </p>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
        data-testid="ChevronRightIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
        />
      </svg>
    </div>
  </div>
</button>
`;
